calyx/opt/src/passes_experimental/fsm_builder.rs

```diff
@@ -0,0 +1,276 @@
+use crate::analysis::{IncompleteTransition, StaticSchedule};
+use crate::traversal::{Action, ConstructVisitor, Named, Visitor};
+use calyx_ir::{self as ir, GetAttributes};
+use calyx_utils::CalyxResult;
+const ACYCLIC: ir::Attribute =
+    ir::Attribute::Internal(ir::InternalAttr::ACYCLIC);
+
+pub struct FSMBuilder {}
+
+impl Named for FSMBuilder {
+    fn name() -> &'static str {
+        "fsm-builder"
+    }
+    fn description() -> &'static str {
+        "generates medium fsms in one pass for static and dynamic"
+    }
+}
+
+impl ConstructVisitor for FSMBuilder {
+    fn from(_ctx: &ir::Context) -> CalyxResult<Self> {
+        Ok(FSMBuilder {})
+    }
+    fn clear_data(&mut self) {}
+}
+
+// node that passes into this function is a part of the schedule
+impl StaticSchedule<'_, '_> {
+    fn build_abstract(
+        &mut self,
+        scon: &ir::StaticControl,
+        guard: ir::Guard<ir::Nothing>,
+        mut transitions_to_curr: Vec<IncompleteTransition>,
```

Comment:
In particular, it is a bit odd that we both _take a mutable vector_ of this type and return another vector of this type. So it seems important to describe exactly what these two vectors do.